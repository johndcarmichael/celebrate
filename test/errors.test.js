/* eslint-env jest */
const expect = require('expect');
const { random } = require('faker');
const {
  celebrate,
  Joi,
  errors,
} = require('../lib');

describe('errors()', () => {
  it('responds with a joi error from celebrate middleware', () => {
    expect.assertions(3);
    const middleware = celebrate({
      query: {
        role: Joi.number().integer().min(4),
      },
    });
    const handler = errors();
    const next = jest.fn();
    const res = {
      status(statusCode) {
        expect(statusCode).toBe(400);
        return {
          send(err) {
            expect(err).toMatchSnapshot();
            expect(next).not.toHaveBeenCalled();
          },
        };
      },
    };

    middleware({
      query: {
        role: random.number({ min: 0, max: 3 }),
      },
      method: 'GET',
    }, null, (err) => {
      handler(err, undefined, res, next);
    });
  });

  it('passes the error through next if not a joi error from celebrate middleware', () => {
    const handler = errors();
    const res = {
      status() {
        throw Error('status called');
      },
    };

    const schema = Joi.object({
      role: Joi.number().integer().min(4),
    });

    const { error } = schema.validate({
      role: random.word(),
    });
    handler(error, undefined, res, (e) => {
      expect(e).toEqual(error);
    });
  });

  it('only includes key values if joi returns details', () => {
    expect.assertions(3);
    const middleware = celebrate({
      body: {
        first: Joi.string().required(),
      },
    });
    const handler = errors();
    const next = jest.fn();
    const res = {
      status(statusCode) {
        expect(statusCode).toBe(400);
        return {
          send(err) {
            expect(err).toMatchSnapshot();
            expect(next).not.toHaveBeenCalled();
          },
        };
      },
    };

    middleware({
      body: {
        role: random.word(),
      },
      method: 'POST',
    }, null, (err) => {
      err.joi.details = null; // eslint-disable-line no-param-reassign
      handler(err, undefined, res, next);
    });
  });
});
