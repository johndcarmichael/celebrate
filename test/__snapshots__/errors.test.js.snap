// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`errors() only includes key values if joi returns details 1`] = `
Object {
  "error": "Bad Request",
  "message": undefined,
  "statusCode": 400,
  "validation": Object {
    "keys": Array [],
    "source": "body",
  },
}
`;

exports[`errors() responds with a joi error from celebrate middleware 1`] = `
Object {
  "error": "Bad Request",
  "message": undefined,
  "statusCode": 400,
  "validation": Object {
    "keys": Array [],
    "source": "query",
  },
}
`;
